### ISSUE TITLE
Support variables in test case names

### ISSUE THREAD

**Issue by kanchi240** [robotframework/robotframework#2962]
Here is the test case, which name contains the variable `${amount}`,

*** Variables ***
${amount}         ${5000000}

*** Test Cases ***
amount cannot larger than ${amount}
    no operation

and below is the log, test case name still contains the `${amount}`, could test case name display to `amount cannot larger than 5000000` in the log?
![15371782599106](https://user-images.githubusercontent.com/4908696/45658478-a3ba2380-bb21-11e8-9863-d3efdd6bbccb.png)

**Comment by pekkaklarck** [comment]
Variables in test case names being resolved, when they exist, sounds good to me, but I don't think that really applies to suite names. Anyone interested to implement this? Shouldn't be too hard but I don't habe time for that. Can onviously help and review pull requests.

**Comment by pekkaklarck** [comment]
The proposed change is to some extend backwards incompatible. Not that badly that it couldn't be implemented, but badly enough that such a change can only be done in a major release. Now before RF 3.1 would be a good time.

**Comment by pekkaklarck** [comment]
PR #3193 looks very good but I still need to think is this a functionality we really want. I added this to RF 3.2 scope to make sure this issue isn't forgotten.

**Comment by michalanglart** [comment]
Hi,

I've just seen this issue and while wondering how would it influence RED editor two questions came to my mind:

- how would it influence the way `--test` selector works?
- what will be the test name reported by `start_test`/`end_test` events of listener API?

**Comment by pekkaklarck** [comment]
Good questions @michalanglart! I think it should go like this:

- `--test` is used before starting execution and variables cannot be resolved at that time. It thus could only work with the original name that contains the variable.
- With the listener API v3 we only send both data object and result object to `start_test` and `end_test`. We in general don't want to modify the data object at execution time, so it would have the original name that contains the variable. The result object should then have the final name as that's the one we wanted to show in reports and logs.
- With the listener API v2 we only send `name`. I don't actually know should it be the original or the final name. Does anyone else have any preferences?

Notice also that I still haven't made a decision is this a good feature to add in general or not. This isn't too high priority issue anyway, and want to have more important RF 3.2 features ready before thinking about this more. If others feel strongly about this either way, please vote the original description with :+1:/:-1: or add a separate comment.

**Comment by pekkaklarck** [comment]
The original description also talked about changing suite names based on variables. At the moment when the name is got from the file/directory name, that probably wouldn't work too well as having a file like `${var}.robot` in the file system would have its own issues. A better solution would be making it possible to support `Name` setting in `Settings` (and obviously allow using variables when setting value for `Name`). A separate issue should be submitted about it.

**Comment by michalanglart** [comment]
> I don't actually know should it be the original or the final name. Does anyone else have any preferences?

From our point of view - we don't have to change anything if the name passed to listener remains original. If it would get the final name on the other hand then it would influence couple of functionalities (because there are places where we need to find a test case knowing it's name from the event and suite: namely the debugger, or some actions in a view which presents the execution tree). Having said that: I would prefer the original one but if it will be the resolved one we will adapt.

> This isn't too high priority issue anyway, and want to have more important RF 3.2 features ready before thinking about this more

Sure, we're just looking through the issues assigned with 3.2 milestone.

**Comment by pekkaklarck** [comment]
I was also thinking that perhaps preserving the original name would be better with listener v2. Don't feel that strongly about it, so if someone has compelling reasons why to use the final name that can be considered.

**Comment by kanchi240** [comment]
Hi, all, thanks for your support, let's just leave this issue and implement it in future if more people need it.

**Comment by pekkaklarck** [comment]
We already have PR #3193 implementing this so the main question is do we want this feature in. I don't see problems with it so let's add it. The aforementioned PR just needs to be enhanced to add docs and to test that listeners work like discussed above.

**Comment by pekkaklarck** [comment]
Thanks to PR #3193 by @yamatoRT this has now been implemented. After merging the PR, I then enhanced docs and tests a bit in the above commits.

This change is backwards incompatible due to these two reasons:
- Existing variables are replaced by their values in test case name. Doing that was the actual motivation of this issue, but it may be problematic in some cases. If it is, variables can be escaped like `\${XXX}`.
- Escapes are resolved so that, for example, single `\` is removed and `\n` becomes a newline. If someone has had something like `c:\path\temp` in a test name, the name will in the future contain `c:path<TAB>emp`. Also in this case escaping like `c:\\path\\temp` works.

Due to these changes, also listeners will get different values than they used to if names contain variables or escapes. With listener v2 the `name` is the resolved name, but they also have new attribute `originalname` that contains the original, unresolved name. With listener v3 the `data` object has the unresolved name and the `result` object has the resolved name.
